runelite-client/src/main/java/net/runelite/client/plugins/runecraft/AbyssOverlay.java

```diff
@@ -85,22 +91,67 @@
 	@Override
 	public Dimension render(Graphics2D graphics)
 	{
-		if (!config.showRifts())
+		if (config.showRifts())
+		{
+			LocalPoint localLocation = client.getLocalPlayer().getLocalLocation();
+			for (DecorativeObject object : plugin.getAbyssObjects())
+			{
+				LocalPoint location = object.getLocalLocation();
+				if (localLocation.distanceTo(location) <= MAX_DISTANCE)
+				{
+					renderRifts(graphics, object);
+				}
+			}
+		}
+
+		if (config.hightlightDarkMage())
 		{
-			return null;
+			highlightDarkMage(graphics);
 		}
 
-		LocalPoint localLocation = client.getLocalPlayer().getLocalLocation();
-		for (DecorativeObject object : plugin.getAbyssObjects())
+		return null;
+	}
+
+	private void highlightDarkMage(Graphics2D graphics)
+	{
+		Query inventoryQuery = new InventoryWidgetItemQuery();
+		WidgetItem[] items = queryRunner.runQuery(inventoryQuery);
+
+		boolean isDegradedPouchInInventory = false;
+		for (WidgetItem item : items)
 		{
-			LocalPoint location = object.getLocalLocation();
-			if (localLocation.distanceTo(location) <= MAX_DISTANCE)
+			switch (item.getId())
 			{
-				renderRifts(graphics, object);
+				case ItemID.MEDIUM_POUCH_5511:
+				case ItemID.LARGE_POUCH_5513:
+				case ItemID.GIANT_POUCH_5515:
+					isDegradedPouchInInventory = true;
+					break;
+				default:
+					continue;
 			}
 		}
 
-		return null;
+		if (!isDegradedPouchInInventory)
+		{
+			return;
+		}
+
+		Query darkMageQuery = new NPCQuery().nameContains("Dark mage");
```

Comment:
Just do the queries once per game tick and store the result instead of per frame.

Reply:
@Adam- Can you point me to the method I need to override that is called per tick? Would I implement this in the "Plugin" class and call it from the "Overlay"?

Reply:
Hook to the gametick event, put the queries there, store the result in the plugin, access it from the overlay. Look at some other plugins, many of them do this.